casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -489,24 +489,27 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Capture: ConnectionsCell: Tr
         .toSet
       allDependencies = (missingParents union missingJustifications).toList
       missingDependencies <- allDependencies.filterA(
-                              blockHash =>
-                                dag
-                                  .lookup(blockHash)
-                                  .map(_.isEmpty && !blockBuffer.exists(_.blockHash == blockHash))
-                            )
+        blockHash =>
+          dag
+            .lookup(blockHash)
+            .map(_.isEmpty))
+      missingUnseenDependencies = missingDependencies.filter(blockHash => !blockBuffer.exists(_.blockHash == blockHash))
       _ <- missingDependencies.traverse(hash => handleMissingDependency(hash, b))
+      _ <- missingUnseenDependencies.traverse(hash => requestMissingDependency(hash))
     } yield ()
 
-  private def handleMissingDependency(hash: BlockHash, parentBlock: BlockMessage): F[Unit] =
+  private def handleMissingDependency(hash: BlockHash, childBlock: BlockMessage): F[Unit] =
     for {
       _ <- blockBufferDependencyDagState.modify(
-            blockBufferDependencyDag =>
-              DoublyLinkedDagOperations
-                .add[BlockHash](blockBufferDependencyDag, hash, parentBlock.blockHash)
-          )
-      _ <- CommUtil.sendBlockRequest[F](BlockRequest(Base16.encode(hash.toByteArray), hash))
+        blockBufferDependencyDag =>
+          DoublyLinkedDagOperations
+            .add[BlockHash](blockBufferDependencyDag, hash, childBlock.blockHash)
```

Comment:
because it can be a missing parent of a block that's already been added to blockBuffer.

IOW two blocks can have one parent.

when the "missing blocks" status is being resolved for a block - this block will be added to the `blockBufferDependencyDag` (with its parents) and to the block buffer.

the `blockBufferDependencyDag` is used in reattempt to get the `dependencyFree` blocks.

now if a block is already in `blockBuffer` but it's a parent of another block (which was requested but not yet processed) then it will never be added to `blockBufferDependencyDag` (because it's in the blockBuffer).
This results in the processing getting "stuck" - the blockBuffer holds a block that should be processed (so it's not allowing new blocks via the network) and the internal "catch up" mechanism isn't aware that the second block has a parent that is already available and should be added (hence blocking the catch up process).

Reply:
Oh whoops; I see my mistake... essentially blockBufferDependencyDag represents blocks that aren't in the block dag. And so blocks in the block buffer don't count.